objects.js

```diff
@@ -9524,4 +9525,132 @@ StagePrompterMorph.prototype.mouseClickLeft = function () {
 
 StagePrompterMorph.prototype.accept = function () {
     this.isDone = true;
+}
+
+// CamSnapshotDialogMorph /////////////////////////////////////////////////////////
+
+/*
+    I am a dialog morph that lets users take a snapshot using their webcam
+    and use it as a costume for their sprites or a background for the Stage.
+*/
+
+// CamSnapshotDialogMorph inherits from DialogBoxMorph:
+
+
+CamSnapshotDialogMorph.prototype = new DialogBoxMorph();
+CamSnapshotDialogMorph.prototype.constructor = CamSnapshotDialogMorph;
+CamSnapshotDialogMorph.uber = DialogBoxMorph.prototype;
+
+// CamSnapshotDialogMorph instance creation
+
+function CamSnapshotDialogMorph(ide, sprite, onCancel, onAccept) {
+    this.init(ide, sprite, onCancel, onAccept);
+};
+
+CamSnapshotDialogMorph.prototype.init = function (ide, sprite, onCancel, onAccept) {
+    this.ide = ide;
+    this.sprite = sprite;
+    this.padding = 10;
+
+    this.oncancel = onCancel;
+    this.accept = onAccept;
+
+    this.videoElement = null; // an HTML5 video element
+    this.videoView = new Morph(); // a morph where we'll copy the video contents
+
+    CamSnapshotDialogMorph.uber.init.call(this);
+
+    this.labelString = 'Camera';
+    this.createLabel();
+
+    this.buildContents();
+};
+
+CamSnapshotDialogMorph.prototype.buildContents = function () {
+    var myself = this,
+        stage = this.ide.stage;
+
+    this.videoElement = document.createElement('video');
+    this.videoElement.hidden = true;
+    document.body.appendChild(this.videoElement);
+
+    if (navigator.mediaDevices && navigator.mediaDevices.getUserMedia) {
+        navigator.mediaDevices.getUserMedia({ video: true })
+            .then(function (stream) {
```

Comment:
You've probably checked, but `navigator.mediaDevices.getUserMedia` is always defined in browsers that support promises?

This effectively blocks all IE's, and requires Chrome and safari w/in the last 3 years. That's probably not too big a deal -- but if `navigator.mediaDevices.getUserMedia` is defined and promises are not then, this seems like a bad situation.

Reply:
Thing is, `mediadevices.getUserMedia` returns a Promise, [as defined by the standard](https://developer.mozilla.org/en-US/docs/Web/API/MediaDevices/getUserMedia). So, nope, IE users are going to be out of luck :(

Should we gray out the button for this case too?

Reply:
IMO, yes. Hopefully, `mediadevices.getUserMedia` is undefined in those case, and no exception would be raised, so it wouldn't be a huge deal, but a grey button is less confusing.

Reply:
For what it's worth, IE doesn't define `mediaDevices.getUserMedia` either.

Why is anybody using that piece of garbage?

Should we gray out the button for IE users? We would need to show a different hint though.

Reply:
Ah true. A different hint would be helpful. We could just add a line or two to the same hint and just say "This doesn't work in IE, and you need a web cam". It semi-publicly shames internet explorer which seems OK. 

Hopefully people are just using Edge now.